### PR TITLE
Incorrect Decorrelation results in wrong plan

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
While attempting to decorrelate the subquery, we were
incorrectly pulling up the join before calculating the window function
of the results of the join. In cases where we have subqueries with
window function and the subquery has outer references we should not
attempt decorrelating it.

There are further optimization which can be done in case of existential
queries but this PR fixes the plan.

Signed-off-by: Jemish Patel <jpatel@pivotal.io>

### CODE REVIEW

File: src/test/regress/expected/qp_correlated_query_optimizer.out [hunk 1]
```diff
@@ -1265,27 +1265,27 @@ select C.j from C where not exists (select max(B.i) from B  where C.i = B.i offs
...
-                           ->  Result  (cost=0.00..431.00 rows=2 width=4)
-                                 ->  Window  (cost=0.00..431.00 rows=2 width=4)
-                                       Order By: b.i
-                                       ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=6 width=4)
-                                             Merge Key: b.i
-                                             ->  Sort  (cost=0.00..431.00 rows=2 width=4)
-                                                   Sort Key: b.i
-                                                   ->  Table Scan on b  (cost=0.00..431.00 rows=2 width=4)
- Settings:  optimizer=on
- Optimizer status: PQO version 2.34.0
```
Comment (vraghavan78) [inline src/test/regress/expected/qp_correlated_query_optimizer.out]: remove the FIXME comment

**Comment by hlinnaka** [comment]
Did the old plan also happen to produce the correct result in this query? Should we add a test case that would demonstrate the incorrect result?

**Comment by jpatel-pivotal** [comment]
@hlinnaka The old plan did not produce the correct results and we added a ORCA test case for it in https://github.com/greenplum-db/gporca/pull/232 

I have also added it here to catch incorrect results.

Thanks.

**Comment by bhuvnesh2703** [comment]
@hlinnaka yes the incorrect plan produced the correct results in this case `not exists`. The join gave some results due to which the tuples from the top query were returned correctly.

**Comment by bhuvnesh2703** [comment]
Was trying to validate the results manually, Interestingly, I don't think so that the results mentioned are correct at all. Even with GPDB planner, 9.5 as well the results are wrong.
Now after the fix we get the same plan and results as planner, but i doubt if they are right.

Consider this:
From the dataset in test case, in the table C has 2 records which match the rank but the results shows only 1. Am i missing something?
```
bhuvneshchaudhary=# select rank() over (order by B.i) from B,C  where C.i = B.i;
 rank
------
    1
    2
    2
    4
    5
    5
(6 rows)

bhuvneshchaudhary=# select i, j from C;
  i  |  j
-----+-----
 288 |   1
   3 |  -1
  32 |  65
  32 |  62
  99 |   7
  78 |  62
   1 | 889
  -1 | 625
   2 |   7
(9 rows)

In C the below 2 records should match the rank for the query:
select C.i, C.j from C where C.i in (select rank() over (order by B.i) from B  where C.i = B.i);
   i  | j
-----+-----
   1 | 889
   2 |   7
But the output is:
select C.i, C.j from C where C.i in (select rank() over (order by B.i) from B  where C.i = B.i);
   i  | j
-----+-----
   1 | 889
```

Plans from ORCA and Planner 
```
bhuvneshchaudhary=# explain select C.j from C where C.i in (select rank() over (order by B.i) from B  where C.i = B.i);
                                                            QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1324037.05 rows=9 width=4)
   ->  Result  (cost=0.00..1324037.05 rows=3 width=4)
         ->  Table Scan on c  (cost=0.00..1324037.05 rows=3 width=4)
               Filter: (subplan)
               SubPlan 1
                 ->  Result  (cost=0.00..431.00 rows=1 width=8)
                       ->  Window  (cost=0.00..431.00 rows=1 width=8)
                             Order By: b.i
                             ->  Sort  (cost=0.00..431.00 rows=1 width=4)
                                   Sort Key: b.i
                                   ->  Result  (cost=0.00..431.00 rows=1 width=4)
                                         Filter: $1 = b.i
                                         ->  Materialize  (cost=0.00..431.00 rows=6 width=4)
                                               ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=6 width=4)
                                                     ->  Table Scan on b  (cost=0.00..431.00 rows=2 width=4)
 Optimizer status: PQO version 2.45.0
(16 rows)

bhuvneshchaudhary=# set optimizer=off;
SET
bhuvneshchaudhary=# explain select C.j from C where C.i in (select rank() over (order by B.i) from B  where C.i = B.i);
                                                  QUERY PLAN
---------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..17.09 rows=5 width=4)
   ->  Seq Scan on c  (cost=0.00..17.09 rows=2 width=4)
         Filter: (subplan)
         SubPlan 1
           ->  Window  (cost=0.00..3.10 rows=1 width=4)
                 Order By: b.i
                 ->  Result  (cost=3.08..3.09 rows=1 width=4)
                       Filter: $0 = b.i
                       ->  Materialize  (cost=3.08..3.09 rows=1 width=4)
                             ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..3.08 rows=1 width=4)
                                   ->  Seq Scan on b  (cost=0.00..3.08 rows=1 width=4)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(13 rows)
```

File: src/test/regress/expected/qp_correlated_query_optimizer.out [hunk 2]
```diff
@@ -150,6 +150,36 @@ select A.i, B.i, C.j from A, B, C where A.j in (select C.j from C where C.j = A.
...
+                                               ->  Materialize  (cost=0.00..431.00 rows=6 width=4)
+                                                     ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=6 width=4)
+                                                           ->  Table Scan on b  (cost=0.00..431.00 rows=2 width=4)
+ Optimizer status: PQO version 2.45.0
+(19 rows)
+
+select C.j from C where C.i in (select rank() over (order by B.i) from B where C.i = B.i) order by C.j;
+  j  
+-----
+ 889
```
Comment (hlinnaka) [inline src/test/regress/expected/qp_correlated_query_optimizer.out]: Shouldn't this return two rows? I believe there are two identical matching rows in C.

Comment (bhuvnesh2703) [reply src/test/regress/expected/qp_correlated_query_optimizer.out]: this is a correlated subquery so should return 1 row only.. i mistakenly made the comment earlier..  :(

**Comment by hlinnaka** [comment]
On 09/15/2017 09:02 AM, Bhuvnesh wrote:
> bhuvnesh2703 commented on this pull request.
> 
> 
> 
>> +                             ->  Window  (cost=0.00..431.00 rows=1 width=8)
> +                                   Order By: b.i
> +                                   ->  Sort  (cost=0.00..431.00 rows=1 width=4)
> +                                         Sort Key: b.i
> +                                         ->  Result  (cost=0.00..431.00 rows=1 width=4)
> +                                               Filter: $1 = b.i
> +                                               ->  Materialize  (cost=0.00..431.00 rows=6 width=4)
> +                                                     ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=6 width=4)
> +                                                           ->  Table Scan on b  (cost=0.00..431.00 rows=2 width=4)
> + Optimizer status: PQO version 2.45.0
> +(19 rows)
> +
> +select C.j from C where C.i in (select rank() over (order by B.i) from B where C.i = B.i) order by C.j;
> +  j
> +-----
> + 889
> 
> this is a correlated subquery so should return 1 row only.. i mistakenly made the comment earlier..  :(

Oh, ok. I got confused while testing myself, I was sure that there were 
two ((1), (889)) rows in table C, but there is only one. Sorry for the 
noise.

- Heikki

**Comment by bhuvnesh2703** [comment]
Thanks @hlinnaka, regarding the plan change.. but correct results, since its a `not_exists` it did not get impacted with the results changes due to incorrect calculation of window function. The other query covers the correctness of results, so i believe we are good with these tests.

**Pull request merged**